tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp

```diff
@@ -0,0 +1,174 @@
+#define LINALG_TPLIMPLEMENTATIONS_INCLUDE_EXPERIMENTAL___P1673_BITS_KOKKOSKERNELS_BLAS2_MATRIX_RANK_1_UPDATE_HPP_
+
+#include <complex>
+
+namespace KokkosKernelsSTD {
+
+namespace Impl {
+
+// manages parallel execution of independent action
+// called like action(i, j) for each matrix element A(i, j)
+template <typename ExecSpace, typename MatrixType>
+class ParallelMatrixVisitor {
+public:
+  KOKKOS_INLINE_FUNCTION ParallelMatrixVisitor(ExecSpace &&exec_in, MatrixType &A_in):
+    exec(exec_in), A(A_in), ext0(A.extent(0)), ext1(A.extent(1))
+  {}
+
+  template <typename ActionType>
+  KOKKOS_INLINE_FUNCTION
+  void for_each_matrix_element(ActionType action) {
+    if (ext0 > ext1) { // parallel rows
+      Kokkos::parallel_for(Kokkos::RangePolicy(exec, 0, ext0),
+        KOKKOS_LAMBDA(const auto i) {
+          using idx_type = std::remove_const_t<decltype(i)>;
+          for (idx_type j = 0; j < ext1; ++j) {
+            action(i, j);
+          }
+        });
+    } else { // parallel columns
+      Kokkos::parallel_for(Kokkos::RangePolicy(exec, 0, ext1),
+        KOKKOS_LAMBDA(const auto j) {
+          using idx_type = std::remove_const_t<decltype(j)>;
+          for (idx_type i = 0; i < ext0; ++i) {
+            action(i, j);
+          }
+        });
+    }
+    exec.fence();
+  }
+
+private:
+  const ExecSpace &exec;
+  MatrixType &A;
+  const size_t ext0;
+  const size_t ext1;
+};
+
+// This version of conj_if_needed() also handles Kokkos::complex<T>
+template <class T>
+KOKKOS_INLINE_FUNCTION
+T conj_if_needed(const T &value)
+{
+  return value;
+};
+
+template <class T>
+KOKKOS_INLINE_FUNCTION
+auto conj_if_needed(const Kokkos::complex<T> &value)
+{
+  return Kokkos::conj(value);
+};
+
+template <class T>
+KOKKOS_INLINE_FUNCTION
+auto conj_if_needed(const std::complex<T> &value)
+{
+  return std::conj(value);
+};
+
+} // namespace Impl
+
+// Nonsymmetric non-conjugated rank-1 update
+// Performs BLAS xGER/xGERU (for real/complex types) A[i,j] += x[i] * y[j]
+
+template<class ExecSpace,
+         class ElementType_x,
+         std::experimental::extents<>::size_type ext_x,
+         class Layout_x,
+         class ElementType_y,
+         std::experimental::extents<>::size_type ext_y,
+         class Layout_y,
+         class ElementType_A,
+         std::experimental::extents<>::size_type numRows_A,
+         std::experimental::extents<>::size_type numCols_A,
+         class Layout_A>
+void matrix_rank_1_update(kokkos_exec<ExecSpace> &&/* exec */,
+  std::experimental::mdspan<ElementType_x, std::experimental::extents<ext_x>, Layout_x,
+    std::experimental::default_accessor<ElementType_x>> x,
+  std::experimental::mdspan<ElementType_y, std::experimental::extents<ext_y>, Layout_y,
+    std::experimental::default_accessor<ElementType_y>> y,
+  std::experimental::mdspan<ElementType_A, std::experimental::extents<numRows_A, numCols_A>, Layout_A,
+    std::experimental::default_accessor<ElementType_A>> A)
+{
+  // constraints
```

Comment:
Constraints are enforced by, well, constraints, concepts, or SFINAE.  Mandates are enforced by `static_assert` or an equivalent.

Reply:
I added the mandates on https://github.com/kokkos/stdBLAS/pull/157/commits/fbc867da270bbcd83c7eb0c31da630e181afd64f, assuming we wish to explicitly check static extents match even if it's strictly implicated by dynamic extent match (checked in the preconditions) - let me know in other case.

I'll remove/replace `static_assert`s for the rank constraints once we conclude the discussion on https://github.com/kokkos/stdBLAS/pull/166#discussion_r846941134.

Reply:
Comment added on https://github.com/kokkos/stdBLAS/pull/157/commits/d935618e4af255e8c7f2e3c5af5b41402480f26c